Work Fake PR: MySQL 8.4.0 - Bug#106645: Slow query log is not logging database/schema name.

jfg956 opened the pull request:
This is a Work Fake PR, with a lot of noisy commit.  Please "thumbs-up" this PR if you think this work is important (other encouragement welcome).

This work is about implementint (not fixing, because it is a feature request) [Bug#106645: Slow query log is not logging database/schema name](https://bugs.mysql.com/bug.php?id=106645).

More details about this work in the [brain dump](https://github.com/jfg956/mysql-server/blob/mysql-8.4.0_bug106645/jfg_brain_dump.md).

This PR is mostly for my own usage to see a clean diff of the whole work.  You can subscribe to it to follow my work, and if you have preliminary comments, feel free to add them in here.

jfg956 committed:
Init branch and brain dump.

jfg956 committed:
Add Fake PR and brain dump.

jfg956 committed:
Scoping this work in brain dump.

jfg956 committed:
Add link to log_slow_extra.

jfg956 committed:
Add link to code for Sys_slow_log_extra in brain dump.

jfg956 committed:
Better brain dump.

jfg956 committed:
Implementation without tests.

jfg956 committed:
Add complexity about how "use ..." is logged in brain-dump.

jfg956 committed:
Fix mistake in test.

(I should have tested the test before comitting)

jfg956 committed:
Improved brain dump.

Add MariaDB output.
Reorg in sections.
Add testing notes.

jfg956 committed:
Update the use... section of the brain dump with revised analysis.

jfg956 committed:
Refactor code, and add db reset.

jfg956 committed:
Update brain dump.

jfg956 committed:
Add idea about logging use all the time,

jfg956 committed:
Expand thought on always logging use...

jfg956 committed:
Refactor for clarity.

jfg956 committed:
Add ntoe about mysqldumpslow broken with administrator command.

jfg956 committed:
Add a bug report for mysqldumpslow.

jfg956 committed:
Complete notes about mysqldumpslow.

jfg956 committed:
log_slow_extra_db enabled by default.

jfg956 committed:
Add tests for the func.

jfg956 committed:
Fix comment in test.

jfg956 committed:
Add tests.

jfg956 committed:
Update brain dump.

grooverdan commented on sql/log.cc:
```diff
@@ -710,10 +714,22 @@ bool File_query_log::write_slow(THD *thd, ulonglong current_utime,
+    } else {
+      log_db_change = false;  /* Do not log db change if db is in the comment line. */
+      db[0] = 0;  /* Reset db triggers logging db change after disabling log_slow_extra_db. */
+      if (thd->db().str) {
```
Honestly I'd just do this bit and not worry about the extra system variable.

adjust scripts/mysqldumpslow.pl.in

When finished add a sample: https://github.com/percona/percona-toolkit/tree/3.x/t/lib/samples/slowlogs

regex:
https://github.com/percona/percona-toolkit/blob/83ba470afe5008276a7656102b8abe0cf40a31e6/lib/SlowLogParser.pm#L45

jfg956 replied:
Thanks for the feedback Daniel.

> I'd just do this bit and not worry about the extra system variable.

If I do this, I might break script parsing the slow log, which I would like to avoid.  With the variable and a default of ON, people will see the brreakage and they will have the liberty to fix their script of disable the feature.  I prefer to give them this libeerty.  Also, with a variable, this patch can be back-ported in 8.0 and 8.4, with a default of OFF.  So I think I will keep it.

> adjust scripts/mysqldumpslow.pl.in

No need, it already swallows everything after `Id: ...`.
- https://github.com/jfg956/mysql-server/blob/mysql-8.4.0/scripts/mysqldumpslow.pl.in#L104

> When finished add a sample [...]

Thanks for bringing this to my attention, I will look into this.

> regex

Thanks for the pointer.  Cheers, J-F

jfg956 replied:
Percona notified: https://perconadev.atlassian.net/browse/PT-2370

jfg956 committed:
Add Daniel Black's comment and minor fix.

jfg956 committed:
Split new tests out of slow_query_log_file_func and in their file (1 of 2).

jfg956 committed:
Split new tests out of slow_query_log_file_func and in their files (2 of 2).

jfg956 committed:
Testing WIP.

jfg956 committed:
Add replication testing, and surprise, to brain dump.

jfg956 committed:
Fix link and add surprise to brain dump.

jfg956 committed:
Improved brain dump, with new surprise.

jfg956 committed:
Fix bug "Db: " instead of "NoDb" on replica.

jfg956 committed:
Update brain dump.

jfg956 committed:
Better comments.

jfg956 committed:
Not doing anything with persisted_variables_extended.

jfg956 committed:
Add Bug and WL nb in tests and putting let in commun section.

jfg956 committed:
Avoid code duplication in check.

jfg956 committed:
Beter comments.

jfg956 committed:
Update brain dump.

jfg956 committed:
Add details about Fake PRs.

jfg956 committed:
Wrap brain dump before RFC PR.

jfg956 committed:
Fix links.

jfg956 committed:
Misc fixes.

jfg956 committed:
Add RFC Fake PR.

jfg956 committed:
Add link to cherry-picked PR.

jfg956 committed:
Remove blogpost TODO from brain dump.

jfg956 committed:
Open Bug #115372.

jfg956 committed:
Weird db name and small fix.

jfg956 committed:
WIP brain dump.

jfg956 committed:
WIP.

jfg956 committed:
Update brain dump with "keeping" multi-line schema bug.

jfg956 committed:
"NoDB" to "Db :" (empty-string db).

jfg956 committed:
Remove buff_len for administrator command.

In my patch for Bug#106645, I am allowing myself to remove the usage of the buff_len
  variable for administrator commandan.  IMHO, it is better cose as it avoids re-using this variable,
  but feel free to revert if you do not like it.

jfg956 committed:
With "Db:", make "Id:" format consisent with "Thread_id:".

In my patch for Bug#106645, I am allowing myself to change the format of Id.
  This avoids a snprintf and is more consistent with Thread_id below (which looks better).
  IMHO, addind Db is the right time to improve on the formatting of this line,
  but feel free to revert if you do not like it.

jfg956 committed:
Inline the declaration of buff_len.

In my patch for Bug#106645, I am allowing myself to inline the declaration of buff_len.
  IMHO, this makes things better as after removing the usage of buff_len further down below,
  here is the only place where it is used, but feel free to revert if you do not like it.

jfg956 committed:
Making usage of end mode clear with better assigments.

In my patch for Bug#106645, I am allowing myself to change the initial assigment of end
  from buff to NULL and assigning it to buff when used.  IMHO it makes more clear
  that there are many independant usages of buff in this function,
  but feel free to revert if you do not like it.

jfg956 committed:
Add links to commit improvements in brain dump.

jfg956 committed:
Fix English errors.

jfg956 committed:
Make test consistent.

jfg956 committed:
Fix indentation.

jfg956 committed:
Fix comment.

jfg956 committed:
Add link to Contrib Fake PR, and remove link to cherry-pick in 8.0.37.

The Cherry-pick details are moved to the brain dump.

jfg956 committed:
Add mergability tests in brain dump.

jfg956 committed:
Add link to PT notifications.

jfg956 commented:
Contributed in below, closing, but keeping the branch.
- https://github.com/jfg956/mysql-server/pull/11
